test/unit/BMath.tree

```diff
@@ -0,0 +1,143 @@
+│   │       └── bo * (1 - (bi / (bi + ai))
+│   └── when equal weights and swap fee is non zero
+│       └── it should return correct value
+│           └── bo * (1 - (bi / (bi + (ai * (1-sf))))
+└── when token weights are unequal
+    ├── when unequal weights and swap fee is zero
+    │   └── it should return correct value
+    │       └── b0 * (1 - (bi / ((bi + ai)))^(wi/wo))
+    └── when unequal weights and swap fee is non zero
+        └── it should return correct value
+            └── b0 * (1 - (bi / ((bi + (ai * (1 - sf)))))^(wi/wo))
+
+BMathTest::calcInGivenOut
+├── when token weight in is zero
+│   └── it should revert // division by zero
+├── when token amount out greater than token balance out
+│   └── it should revert // subtraction underflow
+├── when token amount out equals token balance out
+│   └── it should revert // division by zero
+├── when swap fee greater than BONE
+│   └── it should revert // subtraction underflow
+├── when swap fee equals BONE
+│   └── it should revert // division by zero
+├── when token weight out is zero
+│   └── it should return zero
+├── when token weights are equal
+│   ├── when equal weights and swap fee is zero
+│   │   └── it should return correct value
+│   │       └── bi * ((bo/(bo-ao) - 1)))
+│   └── when equal weights and swap fee is non zero
+│       └── it should return correct value
+│           └── bi * ((bo/(bo-ao) - 1))) / (1 - sf)
+└── when token weights are unequal
+    ├── when unequal weights and swap fee is zero
+    │   └── it should return correct value
+    │       └── bi * (((bo/(bo-ao))^(wo/wi) - 1)))
+    └── when unequal weights and swap fee is non zero
+        └── it should return correct value
+            └── bi * (((bo/(bo-ao))^(wo/wi) - 1))) / (1 - sf)
+
+BMathTest::calcPoolOutGivenSingleIn
+├── when token balance in is zero
+│   └── it should revert // division by zero
+├── when token weight in is zero
+│   └── it should return zero
+├── when total weight is zero
+│   └── it should revert // division by zero
+├── when swap fee is zero
+│   └── it should return correct value
+│       └── ((( ai + bi ) / bi ) ^ (wi/wT)) * pS - pS
+└── when swap fee is non zero
+    └── it should return correct value
+        └── ((( ai * (1 - ((1-(wi/wT))*sf)) + bi) / bi ) ^ (wi/wT)) * pS - pS
+
+BMathTest::calcSingleInGivenPoolOut
+├── when total weight is zero
+│   └── it should revert // division by zero
+├── when swap fee is zero
+│   └── it should return correct value
+│       └── (((pS + ao) / pS) ^ (wT/wi))*bi - bi
+└── when swap fee is non zero
+    └── it should return correct value
+        └── ((((pS + ao) / pS) ^ (wT/wi))*bi - bi) / (1 - ((1 - (wi/wT)) * sf))
+
+BMathTest::calcSingleOutGivenPoolIn
+├── when pool supply is zero
+│   └── it should revert // division by zero
+├── when total weight is zero
+│   └── it should revert // division by zero
+├── when token balance out is zero
+│   └── it should return zero
+├── when swap fee and exit fee are zero
```

Comment:
in retrospect, do you consider all of these testcases would have been easier/cleaner with an array of valid scenarios, like I just did for `bPow`?

Reply:
wdym? we require both to be zero to revert 🤔

Reply:
I commented on the wrong line, sorry. I meant all of the non-reverting cases that `should return correct value`